L1Trigger/L1THGCal/plugins/be_algorithms/HGCalTriggerClusterIdentificationBDT.cc

```diff
@@ -0,0 +1,168 @@
+#include <limits>
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerClusterIdentificationBase.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerTools.h"
+#include "CommonTools/Utils/interface/TMVAEvaluator.h"
+
+
+
+class HGCalTriggerClusterIdentificationBDT : public HGCalTriggerClusterIdentificationBase
+{
+
+  public:
+    class Category
+    {
+      public:
+        Category(float pt_min, float pt_max, float eta_min, float eta_max):
+          pt_min_(pt_min), pt_max_(pt_max), 
+          eta_min_(eta_min), eta_max_(eta_max)
+        {
+        }
+        ~Category(){}
+        bool contains(float pt, float eta) const
+        {
+          bool output = true;
+          if(pt<pt_min_ || pt>=pt_max_) output = false;
+          if(std::abs(eta)<eta_min_ || std::abs(eta)>=eta_max_) output = false;
+          return output;
+        }
+
+        float pt_min() const {return pt_min_;}
+        float pt_max() const {return pt_max_;}
+        float eta_min() const {return eta_min_;}
+        float eta_max() const {return eta_max_;}
+
+      private:
+        float pt_min_ = 0.;
+        float pt_max_ = std::numeric_limits<float>::max();
+        float eta_min_ = 1.5;
+        float eta_max_ = 3.;
+    };
+
+  public:
+
+    HGCalTriggerClusterIdentificationBDT();
+    ~HGCalTriggerClusterIdentificationBDT() override{};
+    void initialize(const edm::ParameterSet& conf) final;
+    float value(const l1t::HGCalMulticluster& cluster) const final;
+    bool decision(const l1t::HGCalMulticluster& cluster) const final;
+
+  private:
+    std::vector<Category> categories_;
+    std::vector<std::unique_ptr<TMVAEvaluator>> bdts_;
+    std::vector<double> working_points_;
+    std::vector<std::string> input_variables_;
+
+    float clusterVariable(const std::string&, const l1t::HGCalMulticluster&) const;
+    int category(float pt, float eta) const;
+
+
+};
+
+DEFINE_EDM_PLUGIN(HGCalTriggerClusterIdentificationFactory,
+    HGCalTriggerClusterIdentificationBDT,
+    "HGCalTriggerClusterIdentificationBDT" );
+
+
+HGCalTriggerClusterIdentificationBDT::
+HGCalTriggerClusterIdentificationBDT():HGCalTriggerClusterIdentificationBase()
+{
+}
+
+void
+HGCalTriggerClusterIdentificationBDT::
+initialize(const edm::ParameterSet& conf)
+{
+  categories_.clear();
```

Comment:
is the idea here that an instance of this object could be re-initialized? Would it make sense just to do the initialization once and then use another object for another set of parameters?